compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -4669,20 +4669,29 @@ static TR::Register *generateMaxMin(TR::Node *node, TR::CodeGenerator *cg, bool
    else
       {
       generateTrg1Src2Instruction(cg, cmp_op, node, condReg, trgReg, src2Reg);
-      if (check_nan)
+      if (is_float)
          {
-         generateConditionalBranchInstruction(cg, TR::InstOpCode::bnun, node, nan_label, condReg);
-         // Move the NaN which is in one of trgReg or src2Reg to trgReg by fadd
-         generateTrg1Src2Instruction(cg, TR::InstOpCode::fadd, node, trgReg, trgReg, src2Reg);
+         // Go to the nan_label for NaN
+         generateConditionalBranchInstruction(cg, TR::InstOpCode::bun, node, nan_label, condReg);
+
+         generateTrg1Src2Instruction(cg, max ? TR::InstOpCode::xsmaxdp : TR::InstOpCode::xsmindp,
+                                     node, trgReg, src1Reg, src2Reg);
          generateLabelInstruction(cg, TR::InstOpCode::b, node, end_label);
+
          generateLabelInstruction(cg, TR::InstOpCode::label, node, nan_label);
+         // Move the NaN which is in one of trgReg or src2Reg to trgReg by fadd
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::fadd, node, trgReg, trgReg, src2Reg);
          }
-      generateConditionalBranchInstruction(cg, max ? TR::InstOpCode::bge : TR::InstOpCode::ble, node, end_label, condReg);
-      generateTrg1Src1Instruction(cg, move_op, node, trgReg, src2Reg);
-
-      dep = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 3, cg->trMemory());
+      else
+         {
+         generateConditionalBranchInstruction(cg, max ? TR::InstOpCode::bge : TR::InstOpCode::ble,
+                                              node, end_label, condReg);
+         generateTrg1Src1Instruction(cg, move_op, node, trgReg, src2Reg);
+         }
+      dep = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 4, cg->trMemory());
       dep->addPostCondition(condReg, TR::RealRegister::NoReg);
       dep->addPostCondition(trgReg, TR::RealRegister::NoReg);
+      dep->addPostCondition(src1Reg, TR::RealRegister::NoReg);
```

Comment:
I missed this bug previously:  trgReg can be the same register as src1Reg in some cases, such that registerAllocator can get into endless loop when the same register appeared twice or more times in a dep condition.